src/io/metadata.cpp

```diff
@@ -489,6 +521,41 @@ void Metadata::SetQuery(const data_size_t* query, data_size_t len) {
   #endif  // USE_CUDA
 }
 
+void Metadata::SetPosition(const data_size_t* positions, data_size_t len) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  // save to nullptr
+  if (positions == nullptr || len == 0) {
+    positions_.clear();
+    num_positions_ = 0;
+    return;
+  }
+  if (num_data_ != len) {
+    Log::Fatal("Length of positions is not same with #data");
+  }
+  if (positions_.empty()) { positions_.resize(num_data_); }
+  num_positions_ = num_data_;
+
+  position_load_from_file_ = false;
+  #ifdef USE_CUDA
+  Log::Fatal("Positions in learning to rank is not supported in CUDA version yet.");
+  #endif  // USE_CUDA
```

Comment:
Could this be moved up further? Think it'd be best right after the first early `return`, so users don't encounter a fatal error when they didn't have `position` data anyway. But that way, we'd avoid doing the work of e.g. `positions_.resize()`.

As a general rule, earlier exceptions are better than later ones because they avoid unnecessary computation, memory usage, etc.

Reply:
Done via [3a2031a](https://github.com/microsoft/LightGBM/pull/5929/commits/3a2031a8678ea32383d33e20fa4e74c2402a912d)